src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -207,10 +210,72 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
                 }
             }
 
+            // We may get a flood of diagnostic notificaitons on the BG.  In order to not overwhelm
+            // the UI thread with work to do, and to avoid excess allocations, we batch up these
+            // notifications, and attempt to process them all at once on the UI thread every 50ms
+            // or so.
+            //
+            // Because we're batching, we can optimize things such that we only store two pieces
+            // of data per provider.  Specifically, if they were removing all diagnostics, and
+            // per-document, the latest diagnostics we've created for it.
+            private static readonly ObjectPool<ProviderIdToBatchedUpdates> s_providerPool = new ObjectPool<ProviderIdToBatchedUpdates>(() => new  ProviderIdToBatchedUpdates());
+            private static readonly ObjectPool<DocumentToUpdateArgs> s_documentPool = new ObjectPool<DocumentToUpdateArgs>(() => new DocumentToUpdateArgs());
+
+            private readonly object _gate = new object();
+            private ProviderIdToBatchedUpdates _idToBatchedUpdates = s_providerPool.Allocate();
+            private Task _updateTask = null;
+
             private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
             {
                 this.AssertIsBackground();
-                RegisterNotification(() => OnDiagnosticsUpdatedOnForeground(e));
+                if (_disposed)
+                {
+                    return;
+                }
+
+                lock (_gate)
+                {
+                    if (!_idToBatchedUpdates.TryGetValue(e.Id, out var batchedUpdates))
+                    {
+                        batchedUpdates = (removeArgs: null, createArgs: s_documentPool.Allocate());
+                    }
+
+                    switch (e.Kind)
+                    {
+                    case DiagnosticsUpdatedKind.DiagnosticsRemoved:
+                        // we're being told about diagnostics going away because a document/project
+                        // was removed.  This supercedes all previous removes and creates for this
+                        // id.
+                        batchedUpdates.removeArgs = e;
+
+                        // Clear all all creates for this provider. 
+                        batchedUpdates.createArgs.Clear();
```

Comment:
❗️ This logic isn't correct: `DiagnosticsRemoved` can be specific to a particular project or document. We need to do two things here:

1. We need to modify data structures as necessary to implement the equivalent of this:

    ```csharp
    batchedUpdates.removeArgs.UnionWith(e);
    ```

2. We need an efficient way to only clear items from `createArgs` that overlap with `e` (and I mean `e` as opposed to `removeArgs`).

Reply:
This is true.  While we remove all diagnostics when we get a DiagnosticsRemoved request:

Reply:
You're right.  While we do remove all diagnostics when we get a DiagnosticsRemoved request:

https://github.com/dotnet/roslyn/blob/1096194309547507cb3f1b8d2fcc3074c8432bab/src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs#L340-L350

We only do this for a removed request that matches our document id.  I need to ensure the latter.

Reply:
#fixed

Reply:
> It's possible the logic is correct based on information which was not obvious to me at the time of review.

Nope.  You're correct.  I missed something.  Removal notifications for other documents would cause me to remove the diagnostics for our document.